### PR TITLE
Document SpinBox Node

### PR DESCRIPTION
**Pull request by zakstam**
Added missing documentation for SpinBox node

### CODE REVIEW

File: doc/classes/SpinBox.xml
Comment (Calinou) [inline doc/classes/SpinBox.xml]: Setter and getter methods shouldn't be spelled out, as they are created automatically. Moreover, the "idiomatic" way to use properties in GDScript as of Godot 3.0 is to access them directly.

File: doc/classes/SpinBox.xml
Comment (Calinou) [inline doc/classes/SpinBox.xml]: To my knowledge, we don't use uppercase for emphasis in the documentation anywhere else, so I think this makes more sense:

```suggestion
		[b]Example:[/b]
```

File: doc/classes/SpinBox.xml
Comment (Calinou) [inline doc/classes/SpinBox.xml]: The "hint" isn't very useful here, I think this is sufficient:

```suggestion
		See the [Range] class for more options over the [SpinBox].
```

**Review by Calinou — DISMISSED** [review]

File: doc/classes/SpinBox.xml
Comment (zaksnet) [reply doc/classes/SpinBox.xml]: > Bold text is now supported in the editor help, so it's a good idea to use it above example codeblocks. 👍

Cool!

Comment (zaksnet) [reply doc/classes/SpinBox.xml]: @Calinou, You think you could add Color tags also?

Comment (Calinou) [reply doc/classes/SpinBox.xml]: @zaksnet I just said it was a possibility, but it's quite involved. It basically involves parsing GDScript code blocks when generating the editor help, inserting placeholder color codes (e.g. `[color={variable_color}]`), then replacing those placeholders to follow the current script editor syntax theme.

Comment (zaksnet) [reply doc/classes/SpinBox.xml]: Sorry, i was talking about `color={variable_color}]` so that i could possibly color a single word, not an entire syntax highlight. Anyway, I hadn't noticed this was implemented already, thanks!

File: doc/classes/SpinBox.xml
Comment (akien-mga) [inline doc/classes/SpinBox.xml]: It's not an id, but the actual LineEdit node instance used in the SpinBox (so that you can call LineEdit-specific methods on it).

File: doc/classes/SpinBox.xml
Comment (akien-mga) [inline doc/classes/SpinBox.xml]: No need to put this in a `_ready()` function, users can put the code block in any method and have it work fine.

File: doc/classes/SpinBox.xml
Comment (akien-mga) [inline doc/classes/SpinBox.xml]: Indentation should be done with 4 spaces, not 2. But if you remove the `func _ready():`, then no indentation is needed.

File: doc/classes/SpinBox.xml
Comment (akien-mga) [inline doc/classes/SpinBox.xml]: There's no reason to add it as child of root, it should simply be `add_child(spin_box)` (so it's a child of the current scene).

File: doc/classes/SpinBox.xml
Comment (akien-mga) [inline doc/classes/SpinBox.xml]: This is incorrect, the `SpinBox` node you instanced was not given the "SpinBox" name, and you put it under root, so you can't access it via `$SpinBox` which would fetch a node named "SpinBox" in the current scene.

Since you already have a reference to the `SpinBox` instance in `spin_box`, you can simply use that instead, so `spin_box.get_line_edit()`.

File: doc/classes/SpinBox.xml
Comment (akien-mga) [inline doc/classes/SpinBox.xml]: No need to save a boolean in a variable before using it, and here it would be better to access the property directly instead of its setter, so:
```
line_edit.context_menu_enabled = false
```

File: doc/classes/SpinBox.xml
Comment (akien-mga) [inline doc/classes/SpinBox.xml]: Same here, no need to save the value in a variable before using it, and direct property access should be favored, so:
```
spin_box.align = LineEdit.ALIGN_RIGHT
```

File: doc/classes/SpinBox.xml
Comment (akien-mga) [inline doc/classes/SpinBox.xml]: I would say "Otherwise, it will be read-only."
You can skip the mention of the default value as #29380 will soon be merged and document all those automatically.

File: doc/classes/SpinBox.xml
Comment (akien-mga) [inline doc/classes/SpinBox.xml]: It would be good to be more explicit on what a prefix and a suffix mean in the content of SpinBox.

File: doc/classes/SpinBox.xml
Comment (zaksnet) [reply doc/classes/SpinBox.xml]: I didn't even know you can do that!

File: doc/classes/SpinBox.xml
Comment (zaksnet) [reply doc/classes/SpinBox.xml]: Isn't it an object ID of the line edit?

File: doc/classes/SpinBox.xml
Comment (akien-mga) [inline doc/classes/SpinBox.xml]: Should be `spin_box.get_line_edit()`. You have to get the LineEdit on the SpinBox *instance*, which is referenced by your `spin_box` variable. If you try the above code, you'll see that it raises an error.

File: doc/classes/SpinBox.xml
Comment (akien-mga) [inline doc/classes/SpinBox.xml]: It's not an object id, it is a reference (a pointer) to the LineEdit instance used by this SpinBox instance. It's the same thing as the `spin_box` variable in your above example, which points to the SpinBox instance.

An object ID is different, it's a number that represents the object's instance. It's not an object, and therefore you couldn't call any methods on it if it were an ID (like you can't call methods on other `int` values).

Comment (zaksnet) [reply doc/classes/SpinBox.xml]: Screenshot_1
Sorry, i might just be misunderstanding. Do you mean its a reference to the object id of the line-edit? in the debugger it says its an object id and also typeof() function says its an object. I'm confused.

Comment (zaksnet) [reply doc/classes/SpinBox.xml]: I can also invoke and LineEdit specific commands using that id.

File: doc/classes/SpinBox.xml
Comment (zaksnet) [reply doc/classes/SpinBox.xml]: missed that :flushed:!

File: doc/classes/SpinBox.xml
Comment (akien-mga) [reply doc/classes/SpinBox.xml]: The debugger shows you that it's an Object, and gives you its ID. But that doesn't mean that the variable is an "Object ID", that's simply its representation in the debugger (the alternative would be to show the LineEdit itself in the debugger, but you can imagine this can quickly because tricky).

LineEdit *is* an Object, like all Nodes. Instanced objects have an Object or instance ID, accessed via `Object.get_instance_id()`, which gives you the identifier by which the engine keeps track of the instance.

So no, you're not using the ID here, you're using a reference to an instanced Object (of the subclass LineEdit, which inherits Object), which has the instance ID 1144.

Comment (zaksnet) [reply doc/classes/SpinBox.xml]: Cool! But:
instance1
instance2

An instance ID ( which can be retrieved using Object.get_instance_id() as you stated) is a simple integer on which you can not invoke any commands. In this case, get_line_edit() method seems to be returning the object itself ( I'm not sure if i should say Object ID or just Object anymore) on which you can freely invoke any method inherited by LineEdit Class. So I think if we say that it returns the Instance ID instead of Object ID it will be misleading (because instance id is just an integer and you can not invoke any commands with it). Please advice.

Comment (akien-mga) [reply doc/classes/SpinBox.xml]: But it's not returning any **ID**. An ID is an integer. What you get here is not an ID, but an **instance** of an Object-derived class (here, a Node, and more specifically a LineEdit node).

I just mentioned the instance ID because this is what you **see** in the debugger as a simple representation of the Object instance pointer, but that's not what you actually get. When you do `get_node("LineEdit")`, you don't get the instance ID, you get the instance itself. Here, it's the same.

Comment (akien-mga) [reply doc/classes/SpinBox.xml]: I pushed an edit clarifying that (it's not that I didn't want to leave it to you, but I want to merge this PR asap before merging another one which will add default values to the classref, and thus introduce conflicts with any unmerged docs PR).

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by akien-mga** [comment]
Cherry-picked for 3.1.2.

**Pull request merged**
